include/swift/AST/AutoDiff.h

```diff
@@ -79,6 +79,123 @@ class AutoDiffParameter {
+  /// Allocates and initializes an empty `AutoDiffParameterIndices` for the
+  /// given `functionType`. `isMethod` specifies whether to treat the function
+  /// as a method.
+  static AutoDiffParameterIndices *create(ASTContext &C,
+                                          AnyFunctionType *functionType,
+                                          bool isMethod);
+
+  bool getIsMethod() const { return isMethod; }
+
+  /// Tests whether none of the parameters are in the set.
+  bool none() const { return indices.none(); }
+
+  /// Adds the indexed parameter to the set. When `isMethod` is not set, the
+  /// indices index into the first parameter list. For example,
+  ///
+  ///   functionType = (A, B, C) -> R
+  ///   paramIndex = 0
+  ///   ==> adds "A" to the set.
+  ///
+  /// When `isMethod` is set, the indices index into the first non-self
+  /// parameter list. For example,
+  ///
+  ///   functionType = (Self) -> (A, B, C) -> R
+  ///   paramIndex = 0
+  ///   ==> adds "A" to the set.
+  ///
+  void addNonSelfParam(unsigned paramIndex);
+
+  /// Adds all the paramaters from the first non-self parameter list to the set.
+  /// For example,
+  ///
+  ///   functionType = (A, B, C) -> R
+  ///   ==> adds "A", B", and "C" to the set.
+  ///
+  ///   functionType = (Self) -> (A, B, C) -> R
+  ///   ==> adds "A", B", and "C" to the set.
+  ///
+  void addAllNonSelfParams();
+
+  /// Adds the self parameter to the set. `isMethod` must be set. For example,
+  ///
+  ///   functionType = (Self) -> (A, B, C) -> R
+  ///   ==> adds "Self" to the set
+  ///
+  void addSelfParam();
+
+  /// Pushes the subset's parameter's types to `paramTypes`, in the order in
+  /// which they appear in the function type. For example,
+  ///
+  ///   functionType = (A, B, C) -> R
+  ///   if "A" and "C" are in the set,
+  ///   ==> pushes {A, C} to `paramTypes`.
+  ///
+  ///   functionType = (Self) -> (A, B, C) -> R
+  ///   if "Self" and "C" are in the set,
+  ///   ==> pushes {Self, C} to `paramTypes`.
+  ///
+  void getSubsetParamTypes(AnyFunctionType *functionType,
+                           SmallVectorImpl<Type> &paramTypes) const;
+
+  /// Returns a bitvector for the SILFunction parameters corresponding to the
+  /// parameters in this set. In particular, this explodes tuples and puts the
+  /// method self parameter at the end. For example,
+  ///
+  ///   functionType = (A, B, C) -> R
+  ///   if "A" and "C" are in the set,
+  ///   ==> returns 101
+  ///   (because the lowered SIL type is (A, B, C) -> R)
+  ///
+  ///   functionType = (Self) -> (A, B, C) -> R
+  ///   if "Self" and "C" are in the set,
+  ///   ==> returns 0011
+  ///   (because the lowered SIL type is (A, B, C, Self) -> R)
+  ///
+  ///   functionType = (A, (B, C), D) -> R
+  ///   if "A" and "(B, C)" are in the set,
+  ///   ==> returns 1110
+  ///   (because the lowered SIL type is (A, B, C, D) -> R)
+  ///
+  const llvm::SmallBitVector getLowered(AnyFunctionType *functionType) const;
```

Comment:
```suggestion
  llvm::SmallBitVector getLowered(AnyFunctionType *functionType) const;
```
`const` isn't really necessary here. It's not a reference anyway. The client may also want to change the result variable.